File: backend/internal/types/canvas.go
package types

import "time"

type Pixel struct {
	X         int    `json:"x"`
	Y         int    `json:"y"`
	Color     string `json:"color"`
	UserID    string `json:"user_id"`
	Username  string `json:"username"`
	Timestamp int64  `json:"timestamp"` // Unix timestamp in milliseconds
}

type User struct {
	ID        string    `json:"id"`
	Username  string    `json:"username"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

type Canvas struct {
	Pixels        [][]string    `json:"pixels,omitempty"`        // Full canvas (for backward compatibility)
	SparsePixels  []SparsePixel `json:"sparse_pixels,omitempty"` // Optimized format
	Size          int           `json:"size"`
	LastUpdate    time.Time     `json:"last_update"`
	Stats         Stats         `json:"stats"`
	RecentChanges []Pixel       `json:"recent_changes"`
}

type SparsePixel struct {
	X     int    `json:"x"`
	Y     int    `json:"y"`
	Color string `json:"color"`
}

type Stats struct {
	TotalPixels     int `json:"total_pixels"`
	UniqueUsers     int `json:"unique_users"`
	PixelsPlacedNow int `json:"pixels_placed_now"`
}

type CooldownInfo struct {
	CanPlace    bool      `json:"can_place"`
	CooldownEnd time.Time `json:"cooldown_end"`
}

type WSMessage struct {
	Type string      `json:"type"`
	Data interface{} `json:"data"`
}

type PlacePixelRequest struct {
	X        int    `json:"x"`
	Y        int    `json:"y"`
	Color    string `json:"color"`
	UserID   string `json:"userId"`
	Username string `json:"username"`
	Size     int    `json:"size"`
}

File: frontend/src/types/canvas.ts
export interface PixelData {
  x: number;
  y: number;
  color: string;
  user_id: string;
  username?: string;
  timestamp: number;
}

export type Pixel = PixelData;

export interface Stats {
  total_pixels?: number;
  totalPixels?: number;
  unique_users?: number;
  uniqueUsers?: number;
  pixels_placed_now?: number;
  pixelsPlacedNow?: number;
}

export interface SparsePixel {
  x: number;
  y: number;
  color: string;
}

export interface Canvas {
  pixels?: string[][]; // Full canvas (for backward compatibility)
  sparse_pixels?: SparsePixel[]; // Optimized format
  sparsePixels?: SparsePixel[]; // Alternative naming
  size: number;
  last_update?: string;
  lastUpdate?: string;
  stats: Stats;
  recent_changes?: PixelData[];
  recentChanges?: PixelData[];
}

export interface CanvasState {
  pixels: string[][];
  lastUpdate: number;
}

export interface User {
  id: string;
  username: string;
  lastPlacement: number;
  isConnected: boolean;
}